### DISCUSSION TITLE
How do I query media items with gatsby-source-wordpress?

### DISCUSSION THREAD

**Discussion by jeffabailey** [gatsbyjs/gatsby#30772] [Help]
This query works perfectly fine in the GraphiQL IDE provided by [the WP GraphQL plugin](https://www.wpgraphql.com/).

```
query LogoQuery {
  mediaItems(where: {title: "logo"}) {
    nodes {
      title
      mediaItemUrl
      mediaDetails {
        height
        width
      }
    }
  }
}
```

I tried dropping the query into a useStaticQuery function, and it fails.

```
const Logo = () => {
  const { allWpMediaItem } = useStaticQuery(
    graphql`
        query LogoQuery {
          mediaItems(where: {title: "logo"}) {
            nodes {
              title
              mediaItemUrl
              mediaDetails {
                height
                width
              }
            }
          }
        }
    `
  )
  console.log('allWpMediaItem', allWpMediaItem)
  // const url = allWpMediaItem.edges[0].node.sourceUrl
  // const altText = allWpMediaItem.edges[0].node.altText
  // const height = allWpMediaItem.edges[0].node.mediaDetails.height
  // const width = allWpMediaItem.edges[0].node.mediaDetails.width

  // console.log("url", url)
  // console.log("altText", altText)
  // console.log("height", height)
  // console.log("width", width)

  return <div>a</div>
}

export default Logo
``` 

Here is the error.

![image](https://user-images.githubusercontent.com/776901/114114458-be964400-9895-11eb-9443-bf1a3e3fe61d.png)

I then used the allWpMediaItem type instead of mediaItems, and the query executes. 

```
import React from "react"
import { Link, useStaticQuery, graphql } from "gatsby"
import "@wordpress/block-library/build-style/style.css"
import Image from "gatsby-image"

const Logo = () => {
  const { allWpMediaItem } = useStaticQuery(
    graphql`
      query LogoQuery {
        allWpMediaItem
        {
          edges {
            node {
              title
              sourceUrl
              altText
            }
          }
        }
      }
    `
  )
  console.log('allWpMediaItem', allWpMediaItem)
  // const url = allWpMediaItem.edges[0].node.sourceUrl
  // const altText = allWpMediaItem.edges[0].node.altText
  // const height = allWpMediaItem.edges[0].node.mediaDetails.height
  // const width = allWpMediaItem.edges[0].node.mediaDetails.width

  // console.log("url", url)
  // console.log("altText", altText)
  // console.log("height", height)
  // console.log("width", width)

  return <div>a</div>
}

export default Logo
```

Unfortunately, this query doesn't return all of the media items and does not return the one titled **logo**.

I also found that taking the second query and running it in the GraphiQL IDE throws an error. The allWpMediaItem type doesn't show up in the console.

![image](https://user-images.githubusercontent.com/776901/114115939-a1af4000-9898-11eb-8975-934bdd453bd0.png)

A few questions come up from this experience.

1. Why can't I copy queries from the GraphiQL IDE and run them in the useStaticQuery function?
2. Why isn't the allWpMediaItem type present in the GraphiQL IDE?
3. How can I get the full set of types used within the useStaticQuery function?
4. Why am I only getting 28 mediaItem records back when I have 189 media items stored in Wordpress? 

I've done quite a bit of google searches and looked at the documentation, but the answers to these questions are still entirely unclear.

**Comment by jeffabailey** [comment]
I found that I can access this URL while running Gatsby to find the types I can query.

http://localhost:8000/__graphql

It is still odd that we can't execute queries that run within the GraphiQL IDE.

**Reply by sanky72** [reply]
Hi @jeffabailey did you find an answer for this - `Why am I only getting 28 mediaItem records back when I have 189 media items stored in Wordpress?`


cc @TylerBarnes  please help I am also facing the same issue. I'm also unable to query all the media items through static query of 
```
query {
    allWpMediaItem(limit: 30) {
      edges {
        node {
          sourceUrl
          localFile {
            publicURL
            childImageSharp {
              fluid(maxWidth: 800) {
                ...GatsbyImageSharpFluid
              }
            }
          }
        }
      }
    }
  }
```

**Reply by JeffersonReal** [reply]
Did anybody make progress with this? I'm experiencing the same confusing results, still unable to find a way to get all WP Media.

I have 25 media items in WP which are all used in published public posts (I read they need to be used in published posts to appear in graphql - makes sense).

In Gatsby (`/___graphql`) `allWpMediaItem{}` I can retrieve **1 media node**.
In GraphiQL IDE `mediaItems (first: 50){}` returns all **25 media nodes**.

I've been building queries from other nodetypes but no matter what my images are not getting fed to Gatsby GraphQL.

**Reply by TylerBarnes** [reply]
@bigupjeff do you have a public WPGraphQL endpoint I can use to reproduce the issue? Hard to say why not all connected media items aren't coming through without having a look at your API

**Reply by JeffersonReal** [reply]
Hey @TylerBarnes, thanks for the quick reply. [This](https://wp-source.bigupweb.uk/graphql) is a fresh WP install with the basic GraphQL plugins + WooCommerce/GraphQL JWT. I can strip all those out for a pure environment if you like, but the problem was the same with only WPGraphQL and Gatsby plugins. Gimme 5 before trying to access, just need to remove auth from that endpoint.

**Reply by JeffersonReal** [reply]
@TylerBarnes auth removed 👍

**Reply by JeffersonReal** [reply]
To reproduce the results that are confusing me:

**My WPGraphQL endpoint (25 nodes)**

```
query NewQuery {
  mediaItems (first: 50) {
    nodes {
      id
    }
  }
}
```

**Gatsby GraphQL (1 node)**
```
query allWpMediaItems {
  allWpMediaItem {
    nodes {
      id
    }
  }
}
```

**Reply by TylerBarnes** [reply]
Those queries typically don't return the same data. gatsby-source-wordpress only fetches media items that are connected to atleast 1 other node. It looks like those other media items aren't connected anywhere. I am seeing something odd though where `WpPost.featuredImage` is returning null when it shouldn't.

I have to jump off now but I'll take another look later this week

**Reply by JeffersonReal** [reply]
Much appreciated thank you.

It was specifically the featured images that were blocking my progress, and more specifically on the **/blog** page node.

"gatsby-source-wordpress only fetches media items that are connected to atleast 1 other node"

All images are used in published posts (some of those are Woo products if that makes a difference and WooGraphQL plugin is installed). If I understand GraphQL correctly, this should mean they are connected to 1 node or more. I'll look into this point in more detail and see if I can debug from this angle.

**Reply by JeffersonReal** [reply]
@TylerBarnes Had another look at this today and made some progress. It seems when I downgrade my WP server from PHP8.0 to PHP7.4, suddenly all my media nodes are pulled into Gatsby as expected. The data is good when I query the API using either PHP8.0/7.4, the nodes only disappear in Gatsby (`___graphql`) so somehow the PHP version is affecting the data available to Gatsby at build time.

**Reply by TylerBarnes** [reply]
Oh wow, good find! Thanks for sharing that, that should help me figure out what's going on. Do you have your WP debug log enabled? It's worth checking to see if there were any errors in there when you were on PHP 8.

**Reply by JeffersonReal** [reply]
I certainly do, annoyingly the wp-gatsby related error remains unchanged (besides the new PHP classification to 'Warning'). I did gain some from wp-graphql-woocommerce on 7.4 though. I don't think these are related as the problem occurs on 8.0.

PHP8.0:
```
[01-Apr-2022 23:06:40 UTC] PHP Warning:  Trying to access array offset on value of type bool in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-gatsby/src/Schema/WPGatsbyWPGraphQLSchemaChanges.php on line 54
```
PHP7.4
```
[01-Apr-2022 23:27:16 UTC] PHP Notice:  Trying to access array offset on value of type bool in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-gatsby/src/Schema/WPGatsbyWPGraphQLSchemaChanges.php on line 54
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
[01-Apr-2022 23:27:17 UTC] PHP Warning:  in_array() expects parameter 2 to be array, string given in /var/www/wp-source.bigupweb.uk/wp-content/plugins/wp-graphql-woocommerce-0.11.0/includes/connection/class-products.php on line 538
```

**Reply by JeffersonReal** [reply]
False PHP alarm! Somehow I must have manifested a false-positive last time I looked at this. The problem is caused by the WP plugin WooGraphQL. Also, I realised I didn't have any inline images set before, so I now know inline images aren't affected, the bug only breaks featured images.

As soon as I disable WooGraphQL I get `featuredImage` nodes. It affects all WP post types, not just Woo products. I'll open an issue once I've collated the facts and stop hijacking this thread. 🤡

**Reply by TylerBarnes** [reply]
Ah, makes sense. fwiw there are only [a few WPGraphQL extensions that are confirmed to work](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/docs/usage-with-popular-wp-graphql-extensions.md#unconfirmed-extensions-might-still-work-but-will-have-problems). I have heard of some users having issues with Woo while for others it seems to work well enough. I think maybe it depends on which features are being used maybe.

**Comment by JeffersonReal** [comment]
@jeffabailey I guess you're past this now, but I just noticed some basic errors in your approach above for anybody else that lands here.

To answer some of your questions:

#### 1. Why can't I copy queries from the GraphiQL IDE and run them in the useStaticQuery function?

Any Gatsby experts are welcome to correct my terminology but basically, the GraphiQL IDE provided by [the WP GraphQL plugin](https://www.wpgraphql.com/) gives you access to the 'internal' WP GraphQL data. Once you build using Gatsby, a new GraphQL schema is constructed for your project from the sourced data (accessible at `/___graphql`). This is why you can't just drop WP GraphiQL IDE queries into Gatsby GraphQL - these data layers are not the same thing!

#### 2. Why isn't the allWpMediaItem type present in the GraphiQL IDE?

Gatsby can build data from any number of sources, so for the sake of avoiding naming clashes, nodes are renamed to be unique in their new context. e.g. `mediaItems` (GraphiQL IDE) becomes `allWpMediaItem` (Gatsby). I've noticed that often, the node structures are also changed, it's not just renaming.

#### 3. How can I get the full set of types used within the useStaticQuery function?

You can use [introspection](https://graphql.org/learn/introspection/).

#### 4. Why am I only getting 28 mediaItem records back when I have 189 media items stored in Wordpress?

Now this is where I am also stuck. Kinda hoping you might have solved this and have the answer for me 🙂

**UPDATE**
My missing media nodes was caused by a conflict between WooGraphQL and Wp-Gatsby, [issue opened here](https://github.com/wp-graphql/wp-graphql-woocommerce/issues/620).

Learn from my mistake and carefully check the basics first. Disable all WP plugins and enable one by one checking for changes in GraphQL data and don't forget, the built-in WP GraphiQL IDE shows you the internal GraphQL schema, the endpoint on your Gatsby site (/___graphql) shows you the Gatsby schema. You need to keep an eye on both to see where your data is getting lost.

**Reply by jeffabailey** [reply]
Hi @bigupjeff , thank you for these responses!

I was able to get past the schema difference confusion, but it sure is confusing for new users.

I'm migrating my Wordpress installation in the near future. I'll keep an eye on the mediaItem record issue. I don't have WooGraphQL installed, but I do have these GraphQL related plugins.

Add WPGraphQL SEO
WPGraphQL JWT Authentication
